modules/shadertools/src/lib/assemble-shaders.js

```diff
@@ -9,13 +9,37 @@ const SHADER_TYPE = {
   [FRAGMENT_SHADER]: 'fragment'
 };
 
+const HOOK_FUNCTIONS = {
+  [VERTEX_SHADER]: {},
+  [FRAGMENT_SHADER]: {}
+};
+
+const MODULE_INJECTIONS = {
+  [VERTEX_SHADER]: {},
+  [FRAGMENT_SHADER]: {}
+};
+
 // Precision prologue to inject before functions are injected in shader
 // TODO - extract any existing prologue in the fragment source and move it up...
 const FRAGMENT_SHADER_PROLOGUE = `\
 precision highp float;
 
 `;
 
+export function setShaderHook(type, opts) {
+  const name = opts.signature.trim().replace(/\(.+/, '');
+  HOOK_FUNCTIONS[type][name] = opts;
```

Comment:
Theoretically a hook function can be overloaded? If we don't allow it, an error should be emitted here.

Reply:
I'm thinking not, because it would become confusing where the module code should be injected and what arguments are available. I'll add the error.